myuw/static/js/card/schedule/visual.js

```diff
@@ -48,22 +48,296 @@ var VisualScheduleCard = {
+                var dates = VisualScheduleCard._get_dates_for_section(section);
+                if(dates[0].isSameOrBefore(week.end_date) && dates[1].isSameOrAfter(week.start_date)){
+                    week.sections.push(section);
+                }
+
+            });
+        });
+
+        return weeks;
+    },
+
+    _get_weeks_from_range: function(range){
+        var start = range[0].week(),
+            end = range[1].week(),
+            weeks = {};
+
+        for (var i = start; i <=end; i++){
+            var year = range[0].year();
+            var week = moment().year(year).week(i);
+            weeks[i] = {
+                "start_date": week.clone().startOf('week'),
+                "end_date": week.clone().endOf('week'),
+                "sections": []
+            };
+        }
+        return weeks;
+
+    },
+
+
+    _get_schedule_range: function(course_data){
+        var earliest,
+            latest;
+        $.each(course_data.sections, function(idx, section){
+            var dates = VisualScheduleCard._get_dates_for_section(section);
+            if (earliest === undefined){
+                earliest = dates[0];
+            } else if(dates[0].isBefore(earliest)){
+                earliest = dates[0];
+            }
+
+            if (latest === undefined){
+                latest = dates[1];
+            } else if(dates[1].isAfter(latest)){
+                latest = dates[1];
+            }
+        });
+        return [earliest, latest];
+    },
+
+    _format_dates: function(schedule_periods){
+        //Formats dates so handlebars helper can parse
+        for (var period_id in schedule_periods){
+            schedule_periods[period_id].start_date = schedule_periods[period_id].start_date.format("YYYY-MM-DD");
+            schedule_periods[period_id].end_date = schedule_periods[period_id].end_date.format("YYYY-MM-DD");
+        }
+        return schedule_periods;
+    },
+
+    _get_finals: function(sections){
+        var finals_periods = {};
+        $.each(sections, function(idx, section){
+            if(section.final_exam !== undefined &&
+                section.final_exam.start_date !== undefined) {
+                VisualScheduleCard._add_to_finals(section, finals_periods);
+            }
+        });
+        return finals_periods;
+    },
+
+    _add_to_finals: function(section, schedule_periods){
+        var week_range = VisualScheduleCard._get_week_range_from_date(section.final_exam.start_date);
+        if ("finals" in schedule_periods){
+            schedule_periods.finals.sections.push(section);
+        } else {
+            schedule_periods.finals =
+            {"start_date": week_range[0].format("YYYY-MM-DD"),
+                "end_date" : week_range[1].format("YYYY-MM-DD"),
+                "sections": [section]
+            };
+        }
+        return schedule_periods;
+    },
+
+
+    _add_to_periods: function(dates, section, periods){
+        if(Object.keys(periods).length === 0){
```

Comment:
Off-term around year change:
latest push fixes and adds test coverage

special case:
Un-needed, added test coverage then removed to verify no changes, in latest push as well